docs/contributing/error-handling.md

```diff
@@ -0,0 +1,366 @@
+
+#### d.IsNewResource() Checks
+
+During resource creation, Terraform CLI expects either a properly applied state for the new resource or an error. To signal proper resource existence, the Terraform Plugin SDK uses an underlying resource identifier (set via `d.SetId(/* some value */)`). If for some reason the resource creation is returned without an error, but also without the resource identifier being set, Terraform CLI will return an error such as:
+
+```
+Error: Provider produced inconsistent result after apply
+
+When applying changes to aws_sns_topic_subscription.sqs,
+provider "registry.terraform.io/hashicorp/aws" produced an unexpected new
+value: Root resource was present, but now absent.
+
+This is a bug in the provider, which should be reported in the provider's own
+issue tracker.
+```
+
+A typical pattern in resource implementations in the `Create`/`CreateContext` function is to `return` the `Read`/`ReadContext` function at the end to fill in the Terraform State for all attributes. Another typical pattern in resource implementations in the `Read`/`ReadContext` function is to remove the resource from the Terraform State if the remote system returns an error or status that indicates the remote resource no longer exists by explicitly calling `d.SetId("")` and returning no error. If the remote system is not strongly read-after-write consistent (eventually consistent), this means the resource creation can return no error and also return no resource state.
+
+To prevent this type of Terraform CLI error, the resource implementation should also check against `d.IsNewResource()` before removing from the Terraform State and returning no error. If that check is `true`, then remote operation error (or one synthesized from the non-existent status) should be returned instead. While adding this check will not fix the resource implementation to handle the eventually consistent nature of the remote system, the error being returned will be less opaque for operators and code maintainers to troubleshoot.
+
+In the Terraform AWS Provider, an initial fix for the Terraform CLI error will typically look like:
+
+```go
+func resourceServiceThingCreate(d *schema.ResourceData, meta interface{}) error {
+    /* ... */
+
+    return resourceServiceThingRead(d, meta)
+}
+
+func resourceServiceThingRead(d *schema.ResourceData, meta interface{}) error {
+    /* ... */
+
+    output, err := conn.DescribeServiceThing(input)
+
+    if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, "ResourceNotFoundException") {
+        log.Printf("[WARN] {Service} {Thing} (%s) not found, removing from state", d.Id())
+        d.SetId("")
+        return nil
+    }
+
+    if err != nil {
+        return fmt.Errorf("error reading {Service} {Thing} (%s): %w", d.Id(), err)
+    }
+
+    /* ... */
+}
+```
+
+Future documentation will show how to properly retry the remote operation for a short period of time until it is successful to remove the error completely.
```

Comment:
Will the "future documentation" replace the _Pull Request Submission and Lifecycle_ document's section on using `tfresource.TimedOut(`) with `resource.Retry()`?

Reply:
Yes 👍